NDTensors/src/lib/Sectors/src/category_product.jl

```diff
@@ -34,78 +45,206 @@ function Base.show(io::IO, s::CategoryProduct)
   return print(io, ")")
 end
 
-category_show(io::IO, k, v) = print(io, v)
-
+category_show(io::IO, ::Int, v) = print(io, v)
 category_show(io::IO, k::Symbol, v) = print(io, "($k=$v,)")
 
+function Base.isless(s1::C, s2::C) where {C<:CategoryProduct}
+  return categories_isless(categories(s1), categories(s2))
+end
+function Base.isless(s1::CategoryProduct, s2::CategoryProduct)
+  return categories_isless(categories(s1), categories(s2))
+end
+
+# =======================================  shared  =========================================
+# there are 2 implementations for CategoryProduct
+# - ordered-like with a Tuple
+# - dictionary-like with a NamedTuple
+
+function categories_fusion_rule(cats1, cats2)
+  diff_cat = CategoryProduct(find_diff(cats1, cats2))
+  shared1, shared2 = find_common(cats1, cats2)
+  fused = map(fusion_rule, values(shared1), values(shared2))
+  return recover_key(typeof(shared1), fused) × diff_cat
+end
+
+# get clean results when mixing implementations
+categories_isequal(::Tuple, ::NamedTuple) = false
+categories_isequal(::NamedTuple, ::Tuple) = false
+
+categories_isless(::NamedTuple, ::Tuple) = throw(ArgumentError("Not implemented"))
+categories_isless(::Tuple, ::NamedTuple) = throw(ArgumentError("Not implemented"))
+
+categories_type(::Type{<:CategoryProduct{T}}) where {T} = T
+
+recover_key(T::Type, t::Tuple{Vararg{AbstractCategory}}) = sector(T, t)
+recover_key(T::Type, c::AbstractCategory) = recover_key(T, (c,))
+recover_key(T::Type, c::CategoryProduct) = recover_key(T, categories(c))
+function recover_key(T::Type, fused::Tuple)
+  # here fused contains at leat one GradedUnitRange
+  g0 = reduce(×, fused)
+  # convention: keep unsorted blocklabels as produced by F order loops in ×
+  new_labels = recover_key.(T, GradedAxes.blocklabels(g0))
+  new_blocklengths =
+    LabelledNumbers.labelled.(GradedAxes.unlabel.(BlockArrays.blocklengths(g0)), new_labels)
+  return GradedAxes.gradedrange(new_blocklengths)
+end
+
+sector(T::Type{<:CategoryProduct}, cats::Tuple) = sector(categories_type(T), cats)
+sector(T::Type, cats::Tuple) = sector(T(cats))  # recover NamedTuple
+sector(args...; kws...) = CategoryProduct(args...; kws...)
```

Comment:
These definitions don't sit right with me, they don't feel principled or that they add much beyond `CategoryProduct`. Can we just use `CategoryProduct` directly in code that currently uses `sector`, and also handle the conversions as needed inline in code that currently uses `sector(T::Type, cats)`?

Reply:
@emstoudenmire curious to hear your thoughts on this design.

Reply:
These functions are used to recover a `CategoryProduct` after doing operations separately on each of its categories. This is used here in `Sectors` for `fusion_rule`, but also in other places, for instance to compute fusion trees (the fusion tree of a `CategoryProduct` is the Kronecker product of the fusion trees of its categories). One typically ends up with a tuple of `AbstractCategory` + the initial type (`CategoryProduct{Tuple}` `CategoryProduct{NamedTuple}`, may be generalized), and need to reconstruct a category with that type.

The `sector` function was defined by @emstoudenmire , I suppose the reason is to have a function instead of relying on implementation details of `CategoryProduct`. I can replace it by calls to `CategoryProduct` if that suits you better.

Reply:
Yes, I found it pretty confusing seeing it in the source code and trying to figure out what it was doing, it is best to remove it for now and just use `CategoryProduct`. I especially don't like the versions that take `Type`s as the first inputs, those seem very arbitrary to me.

If I recall, it was originally meant to be an external convenient constructor for `CategoryProduct` to help simplify the user interface (for example, to avoid introducing `CategoryProduct` directly to users but leave it as an internal implementation detail). We could bring the `sector` function back for that purpose later if we decide that would be helpful, but in a simpler form, say just as an alias for some subset of `CategoryProduct` constructors. However, I still wouldn't want to make use of it in the source code, and I have a feeling it won't be needed for the external interface, for example if we expose users to `×` as the official user interface for constructing product sectors, and maybe create a different non-unicode aliases for that operation, perhaps `product(U1(1), SU2(2))` or `product(A=U1(1), B=SU2(2))`, for users who don't want to use unicode.

Reply:
Please have a look at the latest version where I define `recover_category_product_type`.

Reply:
Thanks for removing the more complicated `sector(...)` constructors, it's definitely an improvement. We can discuss `recover_category_product_type` in another thread.